text/0000-oracles.md

```diff
@@ -0,0 +1,136 @@
+
+### Implementing timeouts
+
+For *request-response* cases there may be a need to handle timeout errors.
+
+In such case *Trigger T1* (see scenario 3 for a reference) after emitting `OracleRequest` to *Oracle O* will also emit another `OracleRequest` to a special *Timeout Oracle* which tracks timeouts. Payload will contain the timeout value.
+
+*Trigger T2* will have more complex activation filter, which also tracks store asset related to *Timeout Oracle*. When execution is started it analyzes triggering event if it got response from *Oracle O* or just awakened by *Timeout Oracle*.
+
+Obviously this solution still relies on some Oracle but just not the original one. As we are in the blockchain nothing can happen without a documented reason or *transaction* to be more precise. So any solution which requires some error checking like that will require another Oracle, just a more trusted one.
+
+### Implementing payment
+
+Payment is not a big deal with centralized solutions, because in that case it can be done off-chain. This section is about automatic decentralized on-chain *Oracles* and payments.
+
+There are a couple of ways to implement that:
+
+1. Pay when the exact Oracle result is retrieved. Such result should contain the amount of money to pay. Probably not very usable solution, because *Oracle* should have a white-list or it can be abused with multiple one-shot accounts.
+2. Basic pre-pay inside trigger. If there will be some change, that *Oracle* transfers it back. The quest is how much should trigger pre-pay? There can be even another *free* `OracleRequest` to retrieve the amount of estimated payment.
+3. Permission to withdraw some amount of money from trigger account. That way *Oracle* can withdraw the exact amount of money it wants for its computations. Also such permission can contain the upper limit of withdrawal.
+
+There are no best solution here. Users should choose whatever they find useful for their exact cases. There can be even more possible ways to process payments.
+
+### Decisions
+
+List all decisions you and other stakeholders made
+
+### Alternatives
+
+1. Make one peer (i.e. *Leader*) directly call some external *API* when trigger submits `EmitOracleRequest` instruction. This is a more on-chain variant and allows us to create integrations with popular oracle providers on our side. However it' not so secure and performant.
+2. Implement our own IBC, make Oracle providers to run their own Iroha 2 based networks and make triggers use this IBC to retrieve data from Oracle blockchains. How exactly Oracle blockchains get data from off-chain world is out of the scope.
+
+### Concerns
+
+There is one unresolved problem:
+
+1. Data signing. Most of the time we want to be sure that the actual computation was done or that the real off-chain server was requested for a specific data. There are some ways to check that, i.e. `TLSNotary`. It's unclear if we should supported it and how.
```

Comment:
~~That is a thing, huh~~

I've been having a similar idea of notarizing TLS sessions for some time, funny to see other people have come up with that

Reply:
Actually I took that idea from "Mastering Ethereum" (or something like that, don't know the original name) book